proposals/2025-02-24_secret_providers.md

```diff
@@ -0,0 +1,277 @@
+
+The second case is that we already have a secret value, but refreshing it has resulted in an error. In this case we should keep the component that uses this secret running with the potentially stale secret, and schedule a retry.
+
+If we are starting up prometheus and do not get any secret values, startup will continue and prometheus will continue to run and retry finding secrets. If there are components that are fully specified, they will run during this time. The idea is that it is better to send partial metrics than no metrics, and the emitted metrics for secrets can alert users if there is a problem with their service providers.
+
+### Secret Refresh and Caching
+
+Prometheus will fetch secrets from configured providers and cache them locally. There is a trade off here between up-to-date credentials with performance considerations and API costs, which users might make different choices on.
+
+* **Local Caching:** Secrets retrieved successfully from a provider are stored in memory
+* **Background Refresh:** A dedicated background process periodically attempts to refresh each configured remote secret from its provider.
+* **Refresh Interval:** A default refresh interval of 1 hour is used, avoiding excessive calls to external secret providers (typical rotation schedules often measured in hours or days). However, this interval should be configurable on a per-provider instance basis. For example:
+```yaml
+scrape_configs:
+  - job_name: 'example'
+    basic_auth:
+      username: 'user'
+      password:
+        vault: # Example provider
+          address: "https://vault.example.com"
+          path: "secret/data/prometheus/example_job"
+          key: "password"
+          refresh_interval: 30m # Override default for this secret
+    # ... other config ...
+```
+* **Caching Behavior on Failure:** If a scheduled background refresh attempt fails (e.g., due to network issues, temporary provider unavailability, invalid credentials after rotation), Prometheus will continue to use the last successfully fetched secret value. This ensures components relying on the secret can continue operating with the last known good credential, preventing outages due to transient refresh problems. Failed refresh attempts will be logged, and the `prometheus_remote_secret_state` metric will reflect the 'stale' state.
+* **Permission Errors:** If a component using a cached secret reports a specific permission or authentication error, this might indicate the cached secret is no longer valid. Hence, an immediate refresh attempt for that specific secret is triggered, bypassing the regular refresh interval.
+* **Not Querying Per-Scrape:** It is explicitly **not** the default behavior to query the secret provider on every use.
+
+
+### Secret rotation
```

Comment:
Is there any example of backend that would benefit from the behaviours defined in this section.

Any example of backend that does not support secret overlapping?

Reply:
> Is there any example of backend that would benefit from the behaviours defined in this section.

This section is less about the secret provider (Vault, AWS) and more about the target system's authentication behaviour during rotation. This can happen if there is any sort of caching during the auth process, or even just a race condition

Specific example:

The target application doesn't query the secret provider directly. Instead, a sidecar pulls the secret (e.g., from AWS Secrets Manager) every 5 minutes and updates a local file/cache the target reads.
Prometheus gets `new_pass` immediately after rotation from AWS.
If the sidecar hasn't refreshed its local cache yet, the target does not knows of `new_pass` so it only accepts `old_pass`.